GuessLookupType improvements

SteveErl opened the pull request:
To improve the accuracy of GuessLookupType(), the
recording type and recording length are now factored in. If the recording type is All, Daily, or Weekly, then it's probably a television show. When there is no episode number or subtitle, the function used to assume it's a movie. But for a Manual Record of a
TV show, there generally isn't an episode number or subtitle,
so this presumption was frequently wrong. To help choose
between movie and TV show, we'll examine the scheduled
length of the recording. If it's less than 90 minutes, we'll guess
it's a television show. Otherwise, we'll guess it's a movie.

Refs #656

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] contribution does not duplicate one of our [existing pull requests](https://github.com/MythTV/mythtv/pulls)
- [x] contribution is in a branch rebased against [master](https://github.com/MythTV/mythtv)
- [x] code compiles successfully without errors
- [x] code follows the [MythTV Coding Standards](https://www.mythtv.org/wiki/Coding_Standards)
- [x] documentation added/updated/removed where necessary
- [x] commits are logically organised and have [good commit messages](https://chris.beams.io/posts/git-commit)

SteveErl commented on mythtv/libs/libmythmetadata/metadatafactory.cpp:
```diff
@@ -654,6 +658,17 @@ LookupType GuessLookupType(ProgramInfo *pginfo)
+    // recording operation is finished, it has been set to
+    // kNotRecording by this time. We'll get the actual
+    // recording type from the RECORD table in the database.
+    RecordingType rt = pginfo->QueryRecordRuleType();
```
Thanks for pointing that out. I didn't realize that the record type information I needed was so easily available. I'll rework the code to use the info from rule->Load(), and eliminate the QueryRecordRuleType() routine which I had added.

If the record type is **Daily** or **Weekly**, then I don't think it could be a movie. In most cases the **All** record type would similarly suggest a TV show, but you have pointed out one scenario in which it might still be a movie. I could declare probable television only for **Daily** or **Weekly**. I'm not thrilled with that because I have a rule generated from EIT where I ask it to record **All** episodes of "The Andy Griffith Show", and I'd like it to guess TV show.

In the kSingleRecord, kOneRecord, or kAllRecord cases, we have nothing to base the decision on. That's why I wanted to consider the length of the recording to give us a hint. "The Andy Griffith Show" is broadcast in 30 minute length, for example. But the length criteria is fallible with unusually short movies or long TV shows. If it's right 90% of the time, is it worth keeping when we're trying to _**guess**_ the type?

The original code said that if it doesn't have a season, an episode, or a subtitle, then it must be a Movie. That's ridiculous. EIT info has no specified season, episode or subtitle where I live. Manual Record rules generally have none of these either. 

Should we return kUnknownVideo for cases where we're not sure?

rcrdnalor commented:
Does this satisfy your intentions ?
```
@@ -670,20 +670,30 @@ LookupType GuessLookupType(ProgramInfo *pginfo)
     else
     {
         // Before committing to something being a movie, we
-        // want to check its rule.  If the rule has a season
-        // or episode number, but the recording doesn't,
-        // and the rec doesn't have a subtitle, this is a
-        // generic recording. If neither the rule nor the
-        // recording have an inetref, season, episode, or
-        // subtitle, it's *probably* a movie.  If it's some
-        // weird combination of both, we've got to try everything.
+        // want to check its rule. If it is a recurring rule for
+        // recordings, this is a tv-show. If we set the rule to 
+        // record all shows, but do not have set a special *search*
+        // option, this is probably an episode of a searies.
+        // If the rule has a season or episode number, but the 
+        // recording doesn't, and the rec doesn't have a subtitle, 
+        // this is a generic recording. 
+        // If neither the rule nor the  recording have an inetref,
+        // season, episode, or subtitle, it's *probably* a movie.
+        // If it's some weird combination of both, we've got to 
+        // try everything.
+        // try everything.
         auto *rule = new RecordingRule();
         rule->m_recordID = pginfo->GetRecordingRuleID();
         rule->Load();
         int ruleepisode = rule->m_episode;
+        RecordingType rulerectype = rule->m_type;
+        RecSearchType rulesearchtype = rule->m_searchType;
         delete rule;
+        RecSearchType rulesearchtype = rule->m_searchType;
         delete rule;
 
-        if (ruleepisode == 0 && pginfo->GetEpisode() == 0 &&
+        if (rulerectype == kDailyRecord || rulerectype == kWeeklyRecord)
+            ret = kProbableTelevision;
+        else if (rulesearchtype == kNoSearch && rulerectype == kAllRecord)
+            ret = kProbableTelevision;
+        else if (ruleepisode == 0 && pginfo->GetEpisode() == 0 &&
             pginfo->GetSubtitle().isEmpty())
             ret = kProbableMovie;
         else if (ruleepisode > 0 && pginfo->GetSubtitle().isEmpty())
```
Please check the comments, too. English is not my native tongue.
Roland

rcrdnalor commented:
BTW, the 
`
rule->Load(); 
`
returns a boolean, we should check the returned value
before accessing  its properties.
Roland

SteveErl commented:
@rcrdnalor : I'm not familiar with RecSearchType. I've looked up the enumeration, but that was not too illuminating. I was hoping for a comment to explain the purpose. So, for the cases we've discussed, how would you expect RecSearchType to be set. You talked about using kAllRecord to record multiple instances of a movie from a fallible satellite source. In that case, what would you expect RecSearchType to be? I talked about requesting a recording all instances of a series based on EIT information through the Program Guide. What would RecSearchType be for this scenario?

I've added a debugging log message in my source to print out the RecSearchType found in GuessLookupType, so I should get some  answers. However, I only have two types of recordings. Manual Record rules to record off of a cable signal, and Program Guide based record rules to record off an over the air antenna. I'm not sure if these two cases will provide me enough data to understand your suggested code.

Just checked my log file. I used the Program Guide to request a single recording of a TV show [**kSingleRecord**], and the RecSearchType was 0 [**kNoSearch**]. I requested recordings of ALL instances of another show [**kAllRecord**], and again the RecSearchType was 0 [**kNoSearch**]. One of my daily Manual Record rules ran [**DailyRecord**], and the RecSearchType was 5 [**kManualSearch**].

Your suggested code works for these cases, which were all TV shows. What would be the RecSearchType be when you record multiple copies of a movie, planning to keep the one with the best quality? It would have to not be kNoSearch.


I've been experimenting with my own alternative, which is to avoid deciding it's a movie when there's no proof.
```
+        // If recording rule is periodic, it's probably a TV show.
+        if ((rulerectype == kDailyRecord) ||
+            (rulerectype == kWeeklyRecord))
+        {
+            LOG(VB_GENERAL, LOG_INFO, QString("GuessLookupType: RecordingType = %1 -> Television")
+                    .arg(rulerectype));
+            return kProbableTelevision;
+        }
...
         if (ruleepisode == 0 && pginfo->GetEpisode() == 0 &&
             pginfo->GetSubtitle().isEmpty())
-            ret = kProbableMovie;
+        {
+            // could be a movie or tv show
+            ret = kUnknownVideo;
+        }
```

Where it used to return kProbableMovie, I now return kUnknownVideo. I think the whole purpose of GuessLookupType() is to save a bit of time on metadata lookup. If we have clear indications that it's a TV show, then we'll only run the TV metadata grabber. If we have clear indications that it's a movie, then we'll only run the movie metadata grabber. In cases where we don't have a clear indication, we should give up and say Unknown. Then the caller should run both TV and movie metadata grabbers and try to pick the best match from the sum of these grabber choices.

The problem I was running into was when no Subtitle, Season, or Episode was set, it was deciding movie. That was a bad presumption. It could still be a TV show or a movie. The proper return should be Unknown.

I've removed the code which was trying to use the recording length to better guess the lookup type. I'm trying to see if I can move that logic into FindBestMatch() and LookerUpper::customEvent(). If we can retrieve the length of each movie or TV show which is a possible match, maybe we can do a better job of picking the best match by comparing those values to the scheduled length of the recording? For example, if there is a movie short which is 10 minutes long, and our recording is 15 minutes, then that seems like a better match than a 60 minute TV show of the same name.

This might help in cases where there are both movies and TV shows with the same name. For example, "Spider-man" or "M*A*S*H". The "M*A*S*H" TV show was 25 minutes long. The "M*A*S*H" movie was 116 minutes. So, if the length of our recording is 2 hours, then the movie metadata is a better match. If the length of our recording is 30 minutes, then the TV show metadata is a better match.

It's requires some fuzzy logic though because we'd have to guess how lengths might be increased by insertion of commercials or how editing content of movies might shorten them. I'm not yet convinced this would work reliably.

rcrdnalor commented:
Returning `kUnknownVideo` does not help, because
https://github.com/MythTV/mythtv/blob/dc4d4f068d892c5b5f0836c0dc0885d5321a3d75/mythtv/libs/libmythmetadata/metadatadownload.cpp#L733
will vote this as a `kProbableMovie`

I will accept the change in 
`LookupType GuessLookupType(ProgramInfo *pginfo)?`
only for that code:
```
        if (rulerectype == kDailyRecord || rulerectype == kWeeklyRecord)
            ret = kProbableTelevision;
```
for the following reasons:
Throughout MythtV the assumption is used that a recording
without a subtitle is meant to be a movie.
If you grep for this, you will find multiple occurrences 
in the MythTV code, the python and perl bindings and mythweb.
Changing one of that occurrences to a different meaning is
not recommended from the architectural point of view.
Changing all occurrences to work like you proposed is
only feasible when implementing a conceptual change.
See the end of the discussion in
http://lists.mythtv.org/pipermail/mythtv-dev/2022-September/079383.html

BTW, MythTV tries hard to fix the EIT for various channels
to move the release information from the subtitle to the description.
Is that what you are looking for? 

Many of the advanced users of MythTV apply local changes to their own 
build and use them. All these users all know these changes will not
be incorporated to mythtv/master, because they hinder
other use-scenarios.

SteveErl commented:
In my development branch I had already modified the handleVideoUndetermined() routine because it wastes time on redundant checks.

```
C symbol: handleVideoUndetermined

  File                 Function                Line
0 metadatadownload.cpp run                     140 list = handleVideoUndetermined(lookup);
1 metadatadownload.cpp handleVideoUndetermined 733 MetadataLookupList
                                                   MetadataDownload::handleVideoUndetermined
                                                   (MetadataLookup *lookup)
2 metadatadownload.h   run                      66 static MetadataLookupList
                                                   handleVideoUndetermined(MetadataLookup*
                                                   lookup);
```
 It only gets called from MetadataDownload::run() where lookup->GetSubtype() has already been checked for kProbableTelevision or kProbableMovie cases. You cannot get to handleVideoUndetermined() in either of those cases, so there's no reason to be testing them again. 

It has already been determined that it's not clearly a TV show or a movie, so we should always run both grabbers and combine the results into one list. The correct code would be
```
MetadataLookupList MetadataDownload::handleVideoUndetermined(MetadataLookup *lookup)
 {
     MetadataLookupList list;

    list.append(handleTelevision(lookup));
    list.append(handleMovie(lookup));
```

>  BTW, MythTV tries hard to fix the EIT for various channels to move the release information from the subtitle to the description. Is that what you are looking for?

Where I'm located, there are 4 local PBS stations where the code in eitfixup.cpp could theoretically help. Unfortunately, the signal for those channels is so weak, that it's not possible for me to tune them in. EIT schedule data is the only thing I can access for those stations. None of the watchable stations with strong signals embed any Season, Episode, or Subtitle in the Description field.

> Throughout MythtV the assumption is used that a recording without a subtitle is meant to be a movie.
 
That bad assumption frequently causes failures, but I won't try to address it under this pull request. I'll set up a new commit which will include the code you approved, and to check the return value of `rule->Load()`, as you suggested.

SteveErl commented:
The latest commit undoes the changes to programinfo.h, programinfo.cpp, and metadatacommon.h. Only metadatafactory.cpp is modified from the original master.

rcrdnalor commented:
You wrote:
> The latest commit undoes the changes to programinfo.h, programinfo.cpp, and metadatacommon.h. Only metadatafactory.cpp is modified from the original master.

Then you should close this pull request and open a new one.
The rationale behind this is to 'not confuse readers of the source code with unnecessary commits'.
Please remember the the last checklist item of the pull request:
```
commits are logically organised and have good commit messages
```
BTW, I was wrong in the assumption of
> BTW, the `rule->Load(); ` returns a boolean, we should check the returned value before accessing its properties. Roland

The code 
https://github.com/MythTV/mythtv/blob/4f01f0ad0cf132798dc61323f0da3c1aa9fbee07/mythtv/libs/libmythmetadata/metadatafactory.cpp#L682
returns already default values. No need to check them twice.

SteveErl commented:
@rcrdnalor : Okay, I'll close this pull and open a new one.

However, I'm confused by your change of mind on rule->Load(). I think you were right in the first place. It doesn't return _default_ values. In recordingrule.cpp we have
```
  53 /** Load a single rule from the recorded table.  Which rule is
  54  *  specified in the member variable m_recordID.
  55  */
  56 bool RecordingRule::Load(bool asTemplate)
  57 {
  58     if (m_recordID <= 0)
  59         return false;
  60 
  61     MSqlQuery query(MSqlQuery::InitCon());
  62     query.prepare("SELECT type, search, " // 00-01
...
  74     query.bindValue(":RECORDID", m_recordID);
  75 
  76     if (!query.exec())
  77     {
  78         MythDB::DBError("SELECT record", query);
  79         return false;
  80     }
  81 
  82     if (!query.next())
  83         return false;
```
Based on the value which is loaded into m_recordID, it is doing a database read. In metadatafactory.cpp, we had
```
         auto *rule = new RecordingRule();
         rule->m_recordID = pginfo->GetRecordingRuleID();
         rule->Load();
```
Even if you're certain that GetRecordingRuleID() wouldn't return a negative number, the SQL database read operation might fail in some way. I still think we should check the return value of Load().

SteveErl commented:
@rcrdnalor : Never mind. I now see that 'new RecordingRule()' does default the values of m_episode and m_type the same way that I was manually defaulting them. I'll remove the check on the return value from Load().
